### PR TITLE
String Calculator kata covering Test Driven Development and Command Query Seperation

### PR DESCRIPTION
**Pull request by sunnykath**
# Kata Information

## Kata name:
String Calculator Kata

## Kata Description Link
Kata Link: https://github.com/MYOB-Technology/General_Developer/blob/main/katas/kata-string-calculator/kata-string-calculator.md

# Learning Outcomes

### Level 3
The basic TDD cycle is followed where in general each time new functionality is added to the solution an appropriate test is written first.

Note: I strictly followed the red-green-blue(refactor) workflow for this kata. This is reflected in my commits as well.

### Level 2
Can identify and provide examples of command methods and query methods in code. Can identify in code methods that violate command query separation.

Note: I tried to seperate command and queries in my methods. This can be seen in the IDelimiters Interface.

### CODE REVIEW

File: StringCalculator/Delimiters/DelimitersHandler.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+{
+    public class DelimitersHandler : IDelimiters
+    {
+        private const string OptionalArg = "//";
+        private const int CustomDelimiterIndex = 2;
+        private const char StartOfCustomDelimiter = '[';
+        private const string CustomDelimiterRegex = "(?<=\\[).+?(?=\\])";
+        
+        private string[] _delimiters = {",", "\n"};        
+        private string _numbersInStringInput = null!;
```
Comment (robertlim89) [inline StringCalculator/Delimiters/DelimitersHandler.cs]:
```suggestion
        private string _numbersInStringInput = "";
```
Best to avoid setting things to `null` if you can manage it, especially if your code doesn't treat it differently to the default value (empty string in this case).

File: StringCalculatorTest/StringCalculationTest.cs [hunk 2]
```diff
@@ -1,13 +1,169 @@
...
+        
         [Fact]
-        public void Test1()
+        public void Should_Return_A_Number_When_Given_A_String()
+        {
+            // Act
+            var result = _stringCalculator.Add("");
+
+            // Assert
+            Assert.IsType<int>(result);
```
Comment (robertlim89) [inline StringCalculatorTest/StringCalculationTest.cs:26]: Do you also want to write a test that also checks that 0 is returned for an empty string (which was one of the rules) or modify this test to check that? Right now it could return 45613643 and still pass.

File: StringCalculatorTest/StringCalculationTest.cs [hunk 3]
```diff
@@ -1,13 +1,169 @@
...
+        public StringCalculationTests()
+        {
+            var delimitersHandler = new DelimitersHandler();
+            var validator = new NumberValidation();
+            _stringCalculator = new StringCalculator(delimitersHandler, validator);
+        }
+        
         [Fact]
-        public void Test1()
+        public void Should_Return_A_Number_When_Given_A_String()
```
Comment (robertlim89) [inline StringCalculatorTest/StringCalculationTest.cs:20]: minor nitpick, but the normal test naming style at MYOB is more like `ShouldReturnANumber_WhenGivenAString` - which (subjectively) makes it easier to separate the return value from the conditions. 

Probably not that important here though

File: StringCalculatorTest/StringCalculationTest.cs [hunk 4]
```diff
@@ -1,13 +1,169 @@
...
+
+            // Assert
+            Assert.Equal(expectedNumber, actualNumber);
+        }
+        
+        [Theory]
+        [InlineData(3, "1,2")]
+        [InlineData(6, "2,4")]
+        [InlineData(8, "3,5")]
+        public void Should_Return_The_Sum_When_Two_Numbers_Are_Passed_In(int expectedSum, string stringInput)
```
Comment (robertlim89) [inline StringCalculatorTest/StringCalculationTest.cs:56]: Any thoughts on multi-digit number inputs for these test cases?

File: StringCalculatorTest/StringCalculationTest.cs [hunk 5]
```diff
@@ -1,13 +1,169 @@
...
+        {
+            // Act & Assert
+            var exception = Assert.Throws<Exception>(() => _stringCalculator.Add(stringInput));
+            Assert.Contains($"Negatives not allowed: {negativeNumbers}", exception.Message);
+        }
+        
+        [Theory]
+        [InlineData(2, "1000,1001,2")]
+        [InlineData(227, "1111,5,3333,222")]
+        public void Should_Return_The_Sum_Of_Numbers_That_Are_Less_Than_A_Thousand(int expectedSum, string stringInput)
```
Comment (robertlim89) [inline StringCalculatorTest/StringCalculationTest.cs:119]: Might be interesting to see what happens if the sum of your inputs is >1000, e.g. "499,501,2,10005" should be 1002.

**Review by robertlim89 — COMMENTED** [review]
Looks good, code is clean.

I especially like how you've made interfaces and implementations for the key classes of the calculator, which is a habit that will serve you well in more complex projects.

For your test cases, the only thing I want to suggest for future is that the test cases in my experience that'll give you the most bang for buck will be things that are on the edge of what is and isn't allowed by your program (e.g. tests around the min and max number - 0 and 1000 here). Other than that I can clearly see that your tests cover what the kata asks for.

File: StringCalculator/Delimiters/DelimitersHandler.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            else
+            {
+                customDelimiters.Add(stringInput[CustomDelimiterIndex] + "");
+            }
+            
+            customDelimiters.AddRange(_delimiters);
+            _delimiters = customDelimiters.ToArray();
+        }
+    }
+}
```
Comment (Lanzhou-J) [inline StringCalculator/Delimiters/DelimitersHandler.cs:59]:
```suggestion
}

```
Not really related to the learning outcomes but it's a good style, see:
https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log

File: StringCalculatorTest/StringCalculationTest.cs [hunk 7]
```diff
@@ -1,13 +1,180 @@
...
+        
         [Fact]
-        public void Test1()
+        public void Should_Return_Zero_When_Given_An_Empty_String()
+        {
+            // Act
+            var result = _stringCalculator.Add("");
+
+            // Assert
+            Assert.Equal(decimal.Zero, result);
```
Comment (Lanzhou-J) [inline StringCalculatorTest/StringCalculationTest.cs:36]: Just a question: Why use "decimal.Zero" instead of "0" here?

File: StringCalculatorTest/StringCalculationTest.cs [hunk 8]
```diff
@@ -1,13 +1,180 @@
...
+        }
+        
+        [Theory]
+        [InlineData(1)]
+        [InlineData(2)]
+        [InlineData(3)]
+        public void Should_Return_The_Same_Number_That_Was_Passed_In(int expectedNumber)
```
Comment (Lanzhou-J) [inline StringCalculatorTest/StringCalculationTest.cs:43]: A suggestion on test name: `ShouldReturnTheSameNumber_WhenSingleNumberWasPassedIn`
in order to be more consistent

File: StringCalculatorTest/StringCalculationTest.cs [hunk 9]
```diff
@@ -1,13 +1,180 @@
...
+            var actualSum = _stringCalculator.Add(stringInput);
+
+            // Assert
+            Assert.Equal(expectedSum, actualSum);
+        }
+
+        [Theory]
+        [InlineData(3, "//;\n1;2")]
+        [InlineData(8, "//-\n3-5")]
+        [InlineData(7, "//+\n6+1")]
```
Comment (Lanzhou-J) [inline StringCalculatorTest/StringCalculationTest.cs:94]:
```suggestion
        [InlineData(7, "//+\n6+1")]
        [InlineData(10, "//+\n6+1/n3")]
```
Maybe not necessary, but I'll probably add one more case that has a different pattern in order to prove "All existing steps should still be supported".

File: StringCalculatorTest/StringCalculationTest.cs [hunk 10]
```diff
@@ -1,13 +1,180 @@
...
+            
+            // Assert
+            Assert.Equal(expectedSum, actualSum);
+        }
+
+        [Theory]
+        [InlineData(6, "//[***]\n1***2***3")]
+        [InlineData(8, "//[---]\n3---5")]
+        [InlineData(7, "//[+++]\n6+++1")]
+        public void Should_Use_The_Delimiter_Provided_Of_Any_Length_In_The_Input_To_Separate_The_Numbers(int expectedSum, string stringInput)
```
Comment (Lanzhou-J) [inline StringCalculatorTest/StringCalculationTest.cs:132]: Might be interesting to add logic to handle the case when the delimiters' lengths are not consistent:
e.g.
`[InlineData(7, "//[+++]\n6++1")]`
Currently get:
`System.FormatException: Input string was not in a correct format.`
Maybe can add a test like `ShouldThrowException_WhenDelimitersLengthsAreNotConsistent`

**Comment by Lanzhou-J** [comment]
The code looks good, very clear commit messages showing a detailed TDD approach. I wonder should there be some unit tests to cover the public functions in DelimitersHandler class and NumberValidation class? 
One benefit of using unit tests is that they help to clarify your intentions.

**Comment by sunnykath** [comment]
> The code looks good, very clear commit messages showing a detailed TDD approach. I wonder should there be some unit tests to cover the public functions in DelimitersHandler class and NumberValidation class? One benefit of using unit tests is that they help to clarify your intentions. @Lanzhou-J 

Initially this functionality was all in one file but I then moved it out to different files. Using the rider test coverage tool, it shows my tests do cover all of it. 
image

I think for this kata, my test are more behavioural test currently compared to implementation. Helps the tests be less coupled with the implementation itself. That's why I don't have as many unit tests. Do you recommend I still add them in regardless?

**Review by robertlim89 — APPROVED** [review]
Good work, keep it up! 😄

**Comment by Lanzhou-J** [comment]
> > The code looks good, very clear commit messages showing a detailed TDD approach. I wonder should there be some unit tests to cover the public functions in DelimitersHandler class and NumberValidation class? One benefit of using unit tests is that they help to clarify your intentions. @Lanzhou-J
> 
> Initially this functionality was all in one file but I then moved it out to different files. Using the rider test coverage tool, it shows my tests do cover all of it. image
> 
> I think for this kata, my test are more behavioural test currently compared to implementation. Helps the tests be less coupled with the implementation itself. That's why I don't have as many unit tests. Do you recommend I still add them in regardless?

Good job. For this kata, you don't need to add them if you don't see DelimiterHandler as a "unit" that can be reused.

It seems to me that your public functions in DelimiterHandler class are covered by your component test instead of a unit test so if strictly follow TDD I would probably suggest adding a unit test file for DelimitersHandler, a unit test for NumberValidation class, a unit test for StringCalculator class using mocked functions in IDelimiters and IValidation, and what you have now -- a component test using "real" DelimitersHandler and NumberValidation class to test whether the communication between two different components is done correctly.
